plugins/org.eclipse.xtext/src/org/eclipse/xtext/serializer/analysis/SemanticSequencerNfaProvider.java

```diff
@@ -63,6 +63,20 @@ public SemNfa(ISemState starts, ISemState stops) {
 		}
 
 		@Override
+		public int hashCode() {
+			if (hashCode == -1)
+				hashCode = new NfaUtil().hashCode(this, GET_ASSIGNED_GRAMMAR_ELEMENT);
```

Comment:
I like to keep them in NfaUtil because 
org.eclipse.xtext.util.formallang.Nfa<STATE>
org.eclipse.xtext.util.formallang.Pda<STATE, STACKITEM>
are two very lightweight, generic interfaces with may use-case-specific implementations. To allow this, the generic, reusable code goes into 
org.eclipse.xtext.util.formallang.NfaUtil
org.eclipse.xtext.util.formallang.PdaUtil

Separating use-case-specific-code from generic code to handle NFAs/PDAs has proven to be useful in the past to reduce complexity and enable reuse.